core/src/test/java/org/apache/iceberg/TestRemoveSnapshots.java

```diff
@@ -387,4 +388,163 @@ public void dataFilesCleanup() throws IOException {
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` staged commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    TableMetadata base = readMetadata();
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // `C` commit
+    table.newAppend()
+        .appendFile(FILE_C)
+        .commit();
+
+    List<String> deletedFiles = new ArrayList<>();
+
+    // Expire all commits including dangling staged snapshot.
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireOlderThan(snapshotB.timestampMillis() + 1)
+        .commit();
+
+    // Files should be deleted of dangling staged snapshot
+    snapshotB.addedFiles().forEach(item -> {
+      Assert.assertTrue(deletedFiles.contains(item.path().toString()));
+    });
+  }
+
+  /**
+   * Expire cherry-pick the commit as shown below, when `B` is in table's current state
+   *  Table:
+   *  A - B
+   *   `- C (B) <--this commit
+   */
+  @Test
+  public void testWithCherryPickTableSnapshot() {
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .commit();
+    // pick the snapshot 'B`
+    TableMetadata base = readMetadata();
+    Snapshot snapshotA = base.snapshots().get(0);
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // Move the table back to `A`
+    table.manageSnapshots()
+        .setCurrentSnapshot(snapshotA.snapshotId())
+        .commit();
+
+    // Generate A -> `C (B)`
+    table.manageSnapshots()
+        .cherrypick(snapshotB.snapshotId())
+        .commit();
+    Snapshot snapshotC = readMetadata().currentSnapshot();
+
+    // Move the table back to `B`
+    table.manageSnapshots()
+        .setCurrentSnapshot(snapshotB.snapshotId())
+        .commit();
+    List<String> deletedFiles = new ArrayList<>();
+
+    // Expire `C`
+    table.expireSnapshots()
+        .deleteWith(deletedFiles::add)
+        .expireSnapshotId(snapshotC.snapshotId())
+        .commit();
+
+    // Make sure no dataFiles are deleted for the B and C snapshot
+    Lists.newArrayList(snapshotB, snapshotC).forEach(i -> {
+      i.addedFiles().forEach(item -> {
+        Assert.assertFalse(deletedFiles.contains(item.path().toString()));
+      });
+    });
+  }
+
+  /**
+   * Test on table below, and expiring
+   *  1) Expire `B`
+   *  2) All commit
+   * Table: A - C - D (B)
+   *          ` B (staged)
+   */
+  @Test
+  public void testWithExpiringStagedThenCherrypick() {
+    // `A` commit
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // `B` commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .stageOnly()
+        .commit();
+
+    // pick the snapshot that's staged but not committed
+    TableMetadata base = readMetadata();
+    Snapshot snapshotB = base.snapshots().get(1);
+
+    // `C` commit to let cherry-pick take effect
```

Comment:
It wasn't clear at first why this was needed, but it is to avoid using a fast-forward commit instead of a cherry-pick. Can you update the comment in `RemoveSnapshots` since the case there isn't possible?

Reply:
Yeah, I have also changed another UT, which was not taking care of fast forward.